File: src/engine/powerflow.ts
import { solveLinear } from './linalg';
import { buildNetwork, busInjection, type Network } from './network';
import {
  BASE_MVA,
  BRANCHES,
  BUS_MAP,
  LOAD_QP_RATIO,
} from './systemModel';
import type {
  BranchResult,
  BusResult,
  OperatingState,
  SolveResult,
} from './types';

// ─────────────────────────────────────────────────────────────────────────────
//  Newton-Raphson 전력조류 해석
// ─────────────────────────────────────────────────────────────────────────────

interface NRSolution {
  V: number[];
  th: number[]; // rad
  converged: boolean;
  iterations: number;
  maxMismatch: number;
  /** 반복별 최대 불일치 추이 (iter0, iter1, …) */
  mismatchHistory: { iter: number; max: number; bus: number; kind: 'P' | 'Q' }[];
}

export function newtonRaphson(net: Network, maxIter = 50, tol = 1e-8): NRSolution {
  const { n, slackLocal, type, vSet } = net;

  const V = new Array(n).fill(1.0);
  const th = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    if (type[i] === 'slack' || type[i] === 'pv') V[i] = vSet[i];
  }

  // 미지수 인덱스: 각도(slack 제외), 전압(PQ 만)
  const angleIdx: number[] = [];
  const voltIdx: number[] = [];
  for (let i = 0; i < n; i++) {
    if (i !== slackLocal) angleIdx.push(i);
    if (type[i] === 'pq') voltIdx.push(i);
  }
  const na = angleIdx.length;
  const nv = voltIdx.length;
  const dim = na + nv;

  let converged = dim === 0;
  let iter = 0;
  let maxMismatch = 0;
  const mismatchHistory: NRSolution['mismatchHistory'] = [];

  for (iter = 0; iter < maxIter && !converged; iter++) {
    // 1) 계산 주입 & 불일치
    const Pc = new Array(n).fill(0);
    const Qc = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      const [p, q] = busInjection(net, V, th, i);
      Pc[i] = p;
      Qc[i] = q;
    }

    const mismatch = new Array(dim).fill(0);
    for (let a = 0; a < na; a++) {
      const i = angleIdx[a];
      mismatch[a] = net.pSched[i] - Pc[i];
    }
    for (let v = 0; v < nv; v++) {
      const i = voltIdx[v];
      mismatch[na + v] = net.qSched[i] - Qc[i];
    }

    // 최대 불일치 성분의 위치(어느 모선의 P인지 Q인지) 추적
    let mmMax = 0;
    let mmIdx = 0;
    for (let k = 0; k < dim; k++) {
      if (Math.abs(mismatch[k]) > mmMax) {
        mmMax = Math.abs(mismatch[k]);
        mmIdx = k;
      }
    }
    maxMismatch = mmMax;
    const mmKind: 'P' | 'Q' = mmIdx < na ? 'P' : 'Q';
    const mmLocal = mmIdx < na ? angleIdx[mmIdx] : voltIdx[mmIdx - na];
    mismatchHistory.push({
      iter,
      max: mmMax,
      bus: net.energized[mmLocal],
      kind: mmKind,
    });

    if (maxMismatch < tol) {
      converged = true;
      break;
    }

    // 2) 야코비안 [H N; M L]
    const J = Array.from({ length: dim }, () => new Array(dim).fill(0));
    const Gm = net.G;
    const Bm = net.B;

    const dPdth = (i: number, j: number): number => {
      if (i === j) return -Qc[i] - Bm[i][i] * V[i] * V[i];
      const ang = th[i] - th[j];
      return V[i] * V[j] * (Gm[i][j] * Math.sin(ang) - Bm[i][j] * Math.cos(ang));
    };
    const dPdV = (i: number, j: number): number => {
      if (i === j) return Pc[i] / V[i] + Gm[i][i] * V[i];
      const ang = th[i] - th[j];
      return V[i] * (Gm[i][j] * Math.cos(ang) + Bm[i][j] * Math.sin(ang));
    };
    const dQdth = (i: number, j: number): number => {
      if (i === j) return Pc[i] - Gm[i][i] * V[i] * V[i];
      const ang = th[i] - th[j];
      return -V[i] * V[j] * (Gm[i][j] * Math.cos(ang) + Bm[i][j] * Math.sin(ang));
    };
    const dQdV = (i: number, j: number): number => {
      if (i === j) return Qc[i] / V[i] - Bm[i][i] * V[i];
      const ang = th[i] - th[j];
      return V[i] * (Gm[i][j] * Math.sin(ang) - Bm[i][j] * Math.cos(ang));
    };

    for (let a = 0; a < na; a++) {
      const i = angleIdx[a];
      for (let aa = 0; aa < na; aa++) J[a][aa] = dPdth(i, angleIdx[aa]);
      for (let vv = 0; vv < nv; vv++) J[a][na + vv] = dPdV(i, voltIdx[vv]);
    }
    for (let v = 0; v < nv; v++) {
      const i = voltIdx[v];
      for (let aa = 0; aa < na; aa++) J[na + v][aa] = dQdth(i, angleIdx[aa]);
      for (let vv = 0; vv < nv; vv++) J[na + v][na + vv] = dQdV(i, voltIdx[vv]);
    }

    // 3) 보정량 적용
    let dx: number[];
    try {
      dx = solveLinear(J, mismatch);
    } catch {
      return { V, th, converged: false, iterations: iter + 1, maxMismatch, mismatchHistory };
    }
    for (let a = 0; a < na; a++) th[angleIdx[a]] += dx[a];
    for (let v = 0; v < nv; v++) V[voltIdx[v]] += dx[na + v];
  }

  return { V, th, converged, iterations: iter, maxMismatch, mismatchHistory };
}

/** 통전 선로의 양단 조류(pu) 계산 */
function branchPower(
  br: { g: number; b: number },
  Vi: number,
  Vj: number,
  thi: number,
  thj: number,
): { pij: number; qij: number; pji: number; qji: number } {
  const { g, b } = br;
  const a = thi - thj;
  const cos = Math.cos(a);
  const sin = Math.sin(a);
  const pij = g * Vi * Vi - Vi * Vj * (g * cos + b * sin);
  const qij = -b * Vi * Vi - Vi * Vj * (g * sin - b * cos);
  // to→from (각도 반전)
  const pji = g * Vj * Vj - Vi * Vj * (g * cos - b * sin);
  const qji = -b * Vj * Vj - Vi * Vj * (-g * sin - b * cos);
  return { pij, qij, pji, qji };
}

export function runPowerFlow(state: OperatingState): SolveResult {
  const net = buildNetwork(state);
  const sol = newtonRaphson(net);

  const buses: Record<number, BusResult> = {};
  const branches: Record<number, BranchResult> = {};

  // ── 모선 결과 ──────────────────────────────────────────────
  for (const bus of Object.values(BUS_MAP)) {
    const li = net.local.get(bus.id);
    if (li === undefined || !sol.converged) {
      // 비가압 또는 미수렴 → 정전 처리
      const load =
        bus.type === 'pq' && bus.category === 'load'
          ? state.loads[bus.id] ?? 0
          : 0;
      buses[bus.id] = {
        id: bus.id,
        energized: false,
        vMag: 0,
        vAngle: 0,
        pGen: 0,
        qGen: 0,
        pLoad: li === undefined ? 0 : load,
        qLoad: 0,
        pInj: 0,
        qInj: 0,
      };
      continue;
    }

    const [pCalc, qCalc] = busInjection(net, sol.V, sol.th, li);
    const pInjMW = pCalc * BASE_MVA;
    const qInjMVar = qCalc * BASE_MVA;

    // 부하 모선만 소비전력 보유 (발전기/수집/배전 모선 부하 = 0)
    const pLoad = bus.category === 'load' ? state.loads[bus.id] ?? 0 : 0;
    const qLoad = pLoad * LOAD_QP_RATIO;

    // 순주입 = 발전 − 부하  ⇒  발전 = 순주입 + 부하 (발전기 모선만 유효)
    const isGen = bus.category === 'generator';
    const pGen = isGen ? pInjMW + pLoad : 0;
    const qGen = isGen ? qInjMVar + qLoad : 0;

    buses[bus.id] = {
      id: bus.id,
      energized: true,
      vMag: sol.V[li],
      vAngle: (sol.th[li] * 180) / Math.PI,
      pGen,
      qGen,
      pLoad,
      qLoad,
      pInj: pInjMW,
      qInj: qInjMVar,
    };
  }

  // ── 선로 결과 ──────────────────────────────────────────────
  let anyOverload = false;
  for (const brDef of BRANCHES) {
    const closed = !!state.breakers[brDef.id];
    const netBr = net.branches.find((b) => b.id === brDef.id);
    const energized = closed && netBr !== undefined && sol.converged;

    if (!energized || !netBr) {
      branches[brDef.id] = {
        id: brDef.id,
        closed,
        energized: false,
        pFrom: 0,
        qFrom: 0,
        pTo: 0,
        qTo: 0,
        pLoss: 0,
        qLoss: 0,
        sFrom: 0,
        loadingPct: 0,
        overloaded: false,
        flowDir: 0,
        flowMW: 0,
      };
      continue;
    }

    const i = netBr.fromLocal;
    const j = netBr.toLocal;
    const { pij, qij, pji, qji } = branchPower(
      netBr,
      sol.V[i],
      sol.V[j],
      sol.th[i],
      sol.th[j],
    );
    const pFrom = pij * BASE_MVA;
    const qFrom = qij * BASE_MVA;
    const pTo = pji * BASE_MVA;
    const qTo = qji * BASE_MVA;
    const sFrom = Math.hypot(pFrom, qFrom);
    const sTo = Math.hypot(pTo, qTo);
    const sMax = Math.max(sFrom, sTo);
    const loadingPct = (sMax / brDef.limitMW) * 100;
    const overloaded = loadingPct > 100;
    if (overloaded) anyOverload = true;

    branches[brDef.id] = {
      id: brDef.id,
      closed,
      energized: true,
      pFrom,
      qFrom,
      pTo,
      qTo,
      pLoss: pFrom + pTo,
      qLoss: qFrom + qTo,
      sFrom,
      loadingPct,
      overloaded,
      flowDir: pFrom >= 0 ? 1 : -1,
      flowMW: Math.abs(pFrom),
    };
  }

  // ── 집계 ──────────────────────────────────────────────────
  let totalGenMW = 0;
  let totalLoadMW = 0;
  for (const b of Object.values(buses)) {
    if (!b.energized) continue;
    totalGenMW += b.pGen;
    totalLoadMW += b.pLoad;
  }
  let totalLossMW = 0;
  for (const br of Object.values(branches)) {
    if (br.energized) totalLossMW += br.pLoss;
  }

  return {
    converged: sol.converged,
    iterations: sol.iterations,
    maxMismatch: sol.maxMismatch,
    mismatchHistory: sol.mismatchHistory,
    buses,
    branches,
    totalGenMW,
    totalLoadMW,
    totalLossMW,
    energizedBuses: net.energized,
    islandedBuses: net.islanded,
    anyOverload,
  };
}

File: python/powerflow.py
"""
Newton-Raphson 전력 조류 계산 (Power Flow)
외부 라이브러리 없이 직접 구현
"""

import math
from network import build_network, bus_injection
from linalg import solve_linear
from system_model import BASE_MVA, BRANCHES, LOAD_QP


def newton_raphson(net: dict, max_iter=50, tol=1e-8) -> dict:
    """
    Newton-Raphson 반복으로 V, θ 계산
    반환: {V, th, converged, iterations, maxMismatch, mismatchHistory}
    """
    n          = net['n']
    slack_loc  = net['slackLocal']
    btype      = net['type']
    vSet       = net['vSet']

    # Flat Start 초기화
    V  = [vSet[i] if btype[i] in ('slack','pv') else 1.0 for i in range(n)]
    th = [0.0]*n

    # 미지수 인덱스
    angle_idx = [i for i in range(n) if i != slack_loc]
    volt_idx  = [i for i in range(n) if btype[i] == 'pq']
    na, nv    = len(angle_idx), len(volt_idx)
    dim       = na + nv

    history = []
    converged = (dim == 0)
    max_mm = 0.0

    if dim == 0:
        return {'V': V, 'th': th, 'converged': True,
                'iterations': 0, 'maxMismatch': 0.0, 'mismatchHistory': []}

    for iteration in range(max_iter):
        # ① 주입전력 계산
        Pc = [0.0]*n; Qc = [0.0]*n
        for i in range(n):
            Pc[i], Qc[i] = bus_injection(net, V, th, i)

        # ② mismatch 벡터
        mm = [0.0]*dim
        for a, i in enumerate(angle_idx):
            mm[a] = net['pSched'][i] - Pc[i]
        for v, i in enumerate(volt_idx):
            mm[na+v] = net['qSched'][i] - Qc[i]

        # ③ 최대 불일치 추적
        max_mm = max(abs(x) for x in mm)
        mm_idx = max(range(dim), key=lambda k: abs(mm[k]))
        if mm_idx < na:
            mm_bus  = net['energized'][angle_idx[mm_idx]]
            mm_kind = 'P'
        else:
            mm_bus  = net['energized'][volt_idx[mm_idx-na]]
            mm_kind = 'Q'
        history.append({'iter': iteration, 'max': max_mm,
                        'bus': mm_bus, 'kind': mm_kind})

        if max_mm < tol:
            converged = True
            break

        # ④ 야코비안 구성
        G, B = net['G'], net['B']

        def dPdth(i, j):
            if i == j: return -Qc[i] - B[i][i]*V[i]**2
            a = th[i]-th[j]
            return V[i]*V[j]*(G[i][j]*math.sin(a) - B[i][j]*math.cos(a))

        def dPdV(i, j):
            if i == j: return Pc[i]/V[i] + G[i][i]*V[i]
            a = th[i]-th[j]
            return V[i]*(G[i][j]*math.cos(a) + B[i][j]*math.sin(a))

        def dQdth(i, j):
            if i == j: return Pc[i] - G[i][i]*V[i]**2
            a = th[i]-th[j]
            return -V[i]*V[j]*(G[i][j]*math.cos(a) + B[i][j]*math.sin(a))

        def dQdV(i, j):
            if i == j: return Qc[i]/V[i] - B[i][i]*V[i]
            a = th[i]-th[j]
            return V[i]*(G[i][j]*math.sin(a) - B[i][j]*math.cos(a))

        J = [[0.0]*dim for _ in range(dim)]
        for a, i in enumerate(angle_idx):
            for aa, ii in enumerate(angle_idx): J[a][aa]    = dPdth(i, ii)
            for vv, ii in enumerate(volt_idx):  J[a][na+vv] = dPdV(i, ii)
        for v, i in enumerate(volt_idx):
            for aa, ii in enumerate(angle_idx): J[na+v][aa]    = dQdth(i, ii)
            for vv, ii in enumerate(volt_idx):  J[na+v][na+vv] = dQdV(i, ii)

        # ⑤ 보정량 계산 및 적용
        dx = solve_linear(J, mm)
        for a, i in enumerate(angle_idx): th[i] += dx[a]
        for v, i in enumerate(volt_idx):  V[i]  += dx[na+v]

    return {'V': V, 'th': th, 'converged': converged,
            'iterations': iteration+1, 'maxMismatch': max_mm,
            'mismatchHistory': history}


def branch_power(br, Vi, Vj, thi, thj):
    """선로 양단 조류 계산"""
    g, b = br['g'], br['b']
    a = thi - thj
    c, s = math.cos(a), math.sin(a)
    pij =  g*Vi**2 - Vi*Vj*(g*c + b*s)
    qij = -b*Vi**2 - Vi*Vj*(g*s - b*c)
    pji =  g*Vj**2 - Vi*Vj*(g*c - b*s)
    qji = -b*Vj**2 - Vi*Vj*(-g*s - b*c)
    return pij, qij, pji, qji


def run_power_flow(state: dict) -> dict:
    """조류계산 전체 실행 — 모선·선로 결과 반환"""
    net = build_network(state)
    sol = newton_raphson(net)
    bus_map = {b['id']: b for b in __import__('system_model').BUSES}

    buses = {}
    for bid in bus_map:
        li = net['local'].get(bid)
        bus = bus_map[bid]
        is_load = bus.get('category', '') == 'load' or bid in state['loads']
        p_load = state['loads'].get(bid, 0) if bid in state['loads'] else 0
        q_load = p_load * LOAD_QP

        if li is None or not sol['converged']:
            buses[bid] = {'id': bid, 'energized': False, 'vMag': 0, 'vAngle': 0,
                          'pGen': 0, 'qGen': 0, 'pLoad': p_load, 'qLoad': 0,
                          'pInj': 0, 'qInj': 0}
            continue

        Pc, Qc = bus_injection(net, sol['V'], sol['th'], li)
        pInj = Pc * BASE_MVA
        qInj = Qc * BASE_MVA
        is_gen = bid in (1, 2)
        buses[bid] = {
            'id': bid, 'energized': True,
            'vMag':   sol['V'][li],
            'vAngle': math.degrees(sol['th'][li]),
            'pGen':   pInj + p_load if is_gen else 0,
            'qGen':   qInj + q_load if is_gen else 0,
            'pLoad':  p_load, 'qLoad': q_load,
            'pInj': pInj, 'qInj': qInj,
        }

    branches = {}
    for brDef in BRANCHES:
        closed  = state['breakers'].get(brDef['id'], False)
        net_br  = next((b for b in net['branches'] if b['id'] == brDef['id']), None)
        en      = closed and net_br and sol['converged']
        if not en:
            branches[brDef['id']] = {'id': brDef['id'], 'closed': closed,
                                      'energized': False, 'pFrom': 0, 'qFrom': 0,
                                      'pTo': 0, 'qTo': 0, 'pLoss': 0,
                                      'loadingPct': 0, 'overloaded': False}
            continue
        i, j = net_br['fromLocal'], net_br['toLocal']
        pij, qij, pji, qji = branch_power(
            net_br, sol['V'][i], sol['V'][j], sol['th'][i], sol['th'][j])
        pFrom, qFrom = pij*BASE_MVA, qij*BASE_MVA
        pTo,   qTo   = pji*BASE_MVA, qji*BASE_MVA
        sFrom = math.hypot(pFrom, qFrom)
        loading = sFrom / brDef['limitMW'] * 100
        branches[brDef['id']] = {
            'id': brDef['id'], 'name': brDef['name'], 'closed': closed,
            'energized': True, 'pFrom': pFrom, 'qFrom': qFrom,
            'pTo': pTo, 'qTo': qTo, 'pLoss': pFrom+pTo,
            'loadingPct': loading, 'overloaded': loading > 100,
        }

    total_gen  = sum(b['pGen']  for b in buses.values() if b['energized'])
    total_load = sum(b['pLoad'] for b in buses.values() if b['energized'])
    total_loss = sum(b['pLoss'] for b in branches.values() if b['energized'])

    return {
        'converged':       sol['converged'],
        'iterations':      sol['iterations'],
        'maxMismatch':     sol['maxMismatch'],
        'mismatchHistory': sol['mismatchHistory'],
        'buses':           buses,
        'branches':        branches,
        'totalGenMW':      total_gen,
        'totalLoadMW':     total_load,
        'totalLossMW':     total_loss,
        'energizedBuses':  net['energized'],
        'islandedBuses':   net['islanded'],
    }
